docs(helm): tutorialize helm/README.md to match docker/README.md (#632)
voxel51/fiftyone-teams-app-deploy: helm/README.md

| **On-Demand**  | Infrequent or GPU-heavy jobs where you don't pay for idle workers                       |

FiftyOne Enterprise 2.14+ provides the configuration for a delegated
operator `teams-do-cpu-default` on CPU nodes in **Always-On** mode.
Enable it by setting
`delegatedOperatorDeployments.deployments.teamsDoCpuDefault.enabled` to
`true` in your `values.yaml`:

```yaml
delegatedOperatorDeployments:
  deployments:
    teamsDoCpuDefault:
      enabled: true
```

Running delegated operators on GPU nodes requires additional configuration.
See
[Leveraging GPU Workloads](./docs/configuring-gpu-workloads.md).
We wrote configuration guides for the major cloud-managed Kubernetes services:

- [Amazon Elastic Kubernetes Service (EKS)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods-2)
- [Google Kubernetes Engine (GKE)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods)
- [Azure Kubernetes Service (AKS)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods-1)

To configure **On-Demand** mode for delegated operators, see

> [Configuring the Kubernetes On-Demand Orchestrator](../docs/orchestrators/configuring-kubernetes-orchestrator.md).

Reviewer: ```suggestion
[Using delegatedOperatorJobTemplates for on-demand executors](docs/configuring-delegated-operators.md#using-delegatedoperatorjobtemplates-for-on-demand-executors).
```

On-demand Orchestrators are now auto-registered 🎉

So now, users can follow the much simpler guide of just configuring the jobs/template values themselves, skipping the `fiftyone.management.register_orchestrator()` steps, etc.

Probably should also mention that an admin will need to visit Settings > Orchestrators, click the orchestrator they want to use, and click Refresh. That'll serve as an e2e test of their orc values, and will make the orchestrator available as a delegation target in the correct parts of the UI.

Author: @mo-getter  where should that note live you think

in the main readme or where the main readme links for on-demand orchs?

  - [:compass: Routing Overview (Path-Based Ingress)](#compass-routing-overview-path-based-ingress)
  - [:memo: Notes](#memo-notes)
- [Step 10: Initial CAS Setup](#step-10-initial-cas-setup)
  - [Add First Admin User](#add-first-admin-user)
  - [Enable Auto Join](#enable-auto-join)
- [Step 11: Test End User Login](#step-11-test-end-user-login)
- [Recommended Enhancements](#recommended-enhancements)
  - [:label: Agentic Labeling](#label-agentic-labeling)
  - [:bricks: Custom Plugin Images](#bricks-custom-plugin-images)
- [Upgrades](#upgrades)
- [Known Issues](#known-issues)
- [Advanced Configuration](#advanced-configuration)
- [Validating](#validating)
- [Health Checks and Monitoring](#health-checks-and-monitoring)
- [Values](#values)

<!-- tocstop -->

## :green_book: Prerequisites Skills and Knowledge

A successful, properly secured deployment of FiftyOne Enterprise requires
knowledge of Kubernetes, Helm, MongoDB, DNS, and TLS/SSL certificate management.
See the chart's
[Prerequisites Skills and Knowledge](./fiftyone-teams-app/README.md#prerequisites-skills-and-knowledge)
for the full list.

## :white_check_mark: Technical Requirements

> A successful, properly secured deployment of FiftyOne Enterprise requires:

Reviewer 2: How we can replace the sentence on line 75 so we aren't repeating line 67.

1. Kubernetes Cluster with `kubectl` configured
    1. Use a
       [supported Kubernetes version](https://kubernetes.io/releases/)
       (and matching the kubectl) version `>=1.31-0`
    1. Refer to the Kubernetes
       [Install Tools documentation](https://kubernetes.io/docs/tasks/tools/)
1. Helm version >= 3.14
    1. Refer to the
       [Installing Helm documentation](https://helm.sh/docs/intro/install/)
1. A MongoDB Database that meets FiftyOne's
   [version constraints](https://docs.voxel51.com/user_guide/config.html#using-a-different-mongodb-version).

> 1. A DNS record or records for ingress

Reviewer 2: ```suggestion
1. A DNS record (or records) for ingress
```

> 1. A TLS/SSL certificate or certificates for HTTPS ingress

Reviewer 2: ```suggestion
1. A TLS/SSL certificate (or certificates) for HTTPS ingress
```

1. (optional) An NFS server or `ReadWriteMany` compatible storage medium for
   [delegated operators](./fiftyone-teams-app/README.md#builtin-delegated-operator-orchestrator),
   [plugins](./fiftyone-teams-app/README.md#plugins),
   and

>    [API high-availability](./fiftyone-teams-app/README.md#highly-available-fiftyone-teams-api-deployments)

Reviewer 2: ```suggestion
   [API high availability](./fiftyone-teams-app/README.md#highly-available-fiftyone-teams-api-deployments)
```

See the example [`values.yaml`](./values.yaml).

> [!TIP]
> When rotating the license,
> restart the `teams-cas` and `teams-api` deployments
> so the new values take effect immediately:
>
> ```shell
> kubectl rollout restart deploy \
>   --namespace your-namespace-here \
>   teams-cas \
>   teams-api
> ```

## :file_folder: Step 3: Choose Authentication Mode

FiftyOne Enterprise supports two authentication modes.
Choose the one that matches your deployment:

| Mode       | Use when                                                                   | `values.yaml` setting                          |
| ---------- | -------------------------------------------------------------------------- | ---------------------------------------------- |
| `internal` | Deployment is Air-gapped or Identity Provider is **OpenID Connect (OIDC)** | `casSettings.env.FIFTYONE_AUTH_MODE: internal` |
| `legacy`   | Identity Provider is **SAML**                                              | `casSettings.env.FIFTYONE_AUTH_MODE: legacy`   |

> [!NOTE]
> The example [`values.yaml`](./values.yaml) ships with
> `casSettings.env.FIFTYONE_AUTH_MODE: legacy` by default.

> > Update it if your Identity Provider uses OIDC.

Reviewer 2: ```suggestion
> Set it to `internal` if your Identity Provider uses OIDC.
```

You can refer to the following docs to set up your Identity Provider

> with FiftyOne:

Reviewer 2: ```suggestion
Refer to the following docs to set up your Identity Provider
with FiftyOne:
```

  [Getting Started with Internal Mode documentation](https://docs.voxel51.com/enterprise/pluggable_auth.html#getting-started-with-internal-mode)

## :gear: Step 4: Configure `values.yaml`

Edit your `values.yaml` file (see the example
[`values.yaml`](./values.yaml) in this directory) to set:

- `fiftyoneLicenseSecrets` — the name of the secret created in
  [Step 2](#closed_lock_with_key-step-2-prepare-license-file)
- `secret.fiftyone.mongodbConnectionString` — the MongoDB connection URI from
  [Step 1](#wrench-step-1-set-up-mongodb-database)
- `secret.fiftyone.cookieSecret` — a randomly generated string
- `secret.fiftyone.encryptionKey` — used to encrypt storage credentials.
  See
  [Storage Credentials and `FIFTYONE_ENCRYPTION_KEY`](./fiftyone-teams-app/README.md#storage-credentials-and-fiftyone_encryption_key)
  for how to generate it
- `secret.fiftyone.fiftyoneAuthSecret` — a randomly generated shared secret
  used for CAS
- `casSettings.env.FIFTYONE_AUTH_MODE` — the mode chosen in
  [Step 3](#file_folder-step-3-choose-authentication-mode)
- `teamsAppSettings.dnsName` — your ingress hostname
- `namespace.name` — set to match your target namespace

>    (e.g. `your-namespace-here`). See the note below

Reviewer 2: ```suggestion
   (e.g. `your-namespace-here`).
   See the note below.
```

    for information on how to configure proxies
- Host-based and path-based routing
  - For host-based routing, see
    [Exposing the `teams-api`](./docs/expose-teams-api.md)

For cloud-specific examples of setting up ingress with TLS, see

- [GKE Deployment Guide](./docs/gke-deployment-guide.md)
- [AWS Deployment Guide](./docs/aws-deployment-guide.md)

## Step 10: Initial CAS Setup

In the steps below, `<DNS_NAME>` is the ingress hostname you set for
`teamsAppSettings.dnsName` in [Step 4](#gear-step-4-configure-valuesyaml).

1. Navigate to the CAS Super Admin UI at
   `https://<DNS_NAME>/cas/configurations`
1. In the *API Key* field (upper right corner),
   enter the value of `secret.fiftyone.fiftyoneAuthSecret`
   (from your `values.yaml`) and select *Sign in*

### Add First Admin User

1. Navigate to the *Admins* tab at
   `https://<DNS_NAME>/cas/admins`
1. Select *Add admin*
1. Provide *Name* and *Email*
1. Select *Add*

### Enable Auto Join

1. Navigate to `https://<DNS_NAME>/cas/providers`
1. Select *+ Edit*
1. Select *Allow auto join*
1. Select *Save*

## Step 11: Test End User Login

Verify the deployment's IdP setup by logging in as a regular user

> 1. In a browser, open `https://<DNS_NAME>/cas`

Reviewer 2: The /cas endpoint redirects to `cas/api-doc` on login.

```suggestion
1. In a browser, open `https://<DNS_NAME>`
```

consider these additional enhancements for a production-ready deployment.

### :label: Agentic Labeling

The Agentic Labeler service provides few-shot VLM inference via vLLM.
Configure a dedicated GPU delegated-operator worker
(see [Leveraging GPU Workloads](./docs/configuring-gpu-workloads.md)).
Start the service via the FiftyOne Enterprise UI under *Settings* -> *Services*.
For an overview of builtin services and the service orchestrator, see the
[service orchestrator documentation](../docs/configuring-service-orchestrator.md).

### :bricks: Custom Plugin Images

If your delegated operators or plugins need additional dependencies,
build and use custom plugin container images.
Use `voxel51/fiftyone-teams-cv-full` as the base image
(that includes a full CV/ML environment), and extend with:

- Custom Python packages
- Internal SDKs or models

After you build the container image, tag the image,
and push it to your container image registry.
Override the default container image in `values.yaml`:

```yaml
delegatedOperatorDeployments:
  deployments:
    teamsDoCpuDefault:
      image:
        repository: <YOUR_CONTAINER_REGISTRY>/fiftyone-cv-full-custom

>         tag: v2.19.0

Reviewer 2: ```suggestion
        tag: v2.24.0
```

| `/rpc`               | `teams-api` | RPC API endpoint                     |
| `/file`              | `teams-api` | File import handling                 |
| `/_pymongo`          | `teams-api` | MongoDB requests via SDK             |
| `/health`            | `teams-api` | Health check endpoint                |

### :memo: Notes

- FiftyOne Enterprise supports routing traffic through proxy servers. Please
  refer to the
  [proxy configuration documentation](./docs/configuring-proxies.md) for
  information on how to configure proxies.
- To expose the `teams-api` for host-based routing, see
  [Exposing the `teams-api`](./docs/expose-teams-api.md).
- For a complete, cloud-specific worked example of ingress/TLS setup, see the
  [GKE Deployment Guide](./docs/gke-deployment-guide.md) or the
  [AWS Deployment Guide](./docs/aws-deployment-guide.md).

## Step 10: Identity Provider (IdP) and Authentication (CAS)

FiftyOne Enterprise uses a Central Authentication Service (CAS), introduced
in v1.6, for centralized login, roles, and user management. You chose your
authentication mode in
[Step 3](#file_folder-step-3-choose-authentication-mode).

> The CAS service requires the following in your `values.yaml`:

Reviewer 2: Why are we repeating the things they already configured? How does this help the user?

      enabled: true
```

Running delegated operators on GPU nodes requires additional configuration.
See
[Leveraging GPU Workloads](./docs/configuring-gpu-workloads.md).
We wrote configuration guides for the major cloud-managed Kubernetes services:

- [Amazon Elastic Kubernetes Service (EKS)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods-2)
- [Google Kubernetes Engine (GKE)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods)
- [Azure Kubernetes Service (AKS)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods-1)

To configure **On-Demand** mode for delegated operators, see
[Using delegatedOperatorJobTemplates for on-demand executors](docs/configuring-delegated-operators.md#using-delegatedoperatorjobtemplates-for-on-demand-executors).
On-demand orchestrators are now auto-registered.

> [!NOTE]
> After configuring an on-demand orchestrator, an admin must go to
> **Settings > Orchestrators** in the FiftyOne Enterprise UI, select the
> orchestrator, and click **Refresh**. This end-to-end tests your job
> template values and makes the orchestrator available as a delegation

> > target in the UI.

Reviewer 2: Semantic breaks.

```suggestion
> [!NOTE]
> After configuring an on-demand orchestrator, an admin must
> 
> 1. In the FiftyOne Enterprise UI go to *Settings* -> *Orchestrators*
> 1. Select the orchestrator
> 1. Select *Refresh*
>
> This end-to-end tests your job template values
> and makes the orchestrator available as a delegation target in the UI.
```

operator `teams-do-cpu-default` on CPU nodes in **Always-On** mode.
Enable it by setting
`delegatedOperatorDeployments.deployments.teamsDoCpuDefault.enabled` to
`true` in your `values.yaml`:

```yaml
delegatedOperatorDeployments:
  deployments:
    teamsDoCpuDefault:
      enabled: true
```

Running delegated operators on GPU nodes requires additional configuration.
See
[Leveraging GPU Workloads](./docs/configuring-gpu-workloads.md).
We wrote configuration guides for the major cloud-managed Kubernetes services:

- [Amazon Elastic Kubernetes Service (EKS)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods-2)
- [Google Kubernetes Engine (GKE)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods)
- [Azure Kubernetes Service (AKS)](./docs/configuring-gpu-workloads.md#deploying-gpu-enabled-delegated-operator-pods-1)

To configure **On-Demand** mode for delegated operators, see
[Using delegatedOperatorJobTemplates for on-demand executors](docs/configuring-delegated-operators.md#using-delegatedoperatorjobtemplates-for-on-demand-executors).
On-demand orchestrators are now auto-registered.

> > [!NOTE]
> > After configuring an on-demand orchestrator, an admin must
> > 
> > 1. In the FiftyOne Enterprise UI go to *Settings* -> *Orchestrators*
> > 1. Select the orchestrator
> > 1. Select *Refresh*
> >
> > This end-to-end tests your job template values
> > and makes the orchestrator available as a delegation target in the UI.

Reviewer 2: ```suggestion
> [!NOTE]
> After configuring an on-demand orchestrator, an admin must
>
> 1. In the FiftyOne Enterprise UI go to *Settings* -> *Orchestrators*
> 1. Select the orchestrator
> 1. Select *Refresh*
>
> This end-to-end tests your job template values
> and makes the orchestrator available as a delegation target in the UI.
```
